File: src/3-generators/2-yield-factorial.cs
using static System.Console;
using System.Collections;
using System.Collections.Generic;

static class Program
{
	public static void Main(string[] args)
	{
		var enumerator = new FactorialEnumerator();
		for (int i = 0; i < 10; i++)
		{
			enumerator.MoveNext();
			WriteLine(enumerator.Current);
		}
	}

    static IEnumerator<int> Factorial()
    {
	    var result = 1;
	    var n = 1;
	    while (true)
	    {
		    yield return result;
		    n++;
		    result *= n;
	    }
    }

//	static void F()
//	{
//		yield return 1;
//	}
	
	class FactorialEnumerator : IEnumerator<int>
	{
		private int result = 1;
		private int n = 0;
        
		public bool MoveNext()
		{
			n++;
			result *= n;
			return true;
		}

		public int Current => result;

		public void Reset() => result = 1;
		public void Dispose() {} 
		object IEnumerator.Current => Current;
	} 
}

File: src/3-generators/2-yield-factorial.js
function* foo() {
  yield 1;
  yield 2;
  // ["a", "b", "c"].forEach(function(it){ yield it; })
}

function* factorial() {
  let n = 0;
  let result = 1;
  let skip = 0;
  while (true) {
    if (!skip) {
      skip = yield result;
    } else {
      skip--;
    }
    n++;
    result = result * n;
  }
}

const g = foo();
console.log(g.next());
console.log(g.next());
console.log(g.next());

const f = factorial();
console.log(f.next());
console.log(f.next());
console.log(f.next());
console.log(f.next());
console.log(f.next(10));
